### ISSUE TITLE
Unable to sign ssh certificate

### ISSUE THREAD

**Issue by peyr0l** [Yubico/python-yubihsm#32]
Hello.
I'm struggling to have the ssh certificate signed using sign_ssh_certificate method.

I'm running the following python code:

```
hsm_url = 'http://***:12347/'
hsm_user_auth_pass = '***'
hsm_user_auth_key = 0x4001
sshca_key_id = 0x4012
sshca_template_id = 0x4011

def certify():
    hsm = YubiHsm.connect(hsm_url)
    ses = hsm.create_session_derived(hsm_user_auth_key, hsm_user_auth_pass)
    key = AsymmetricKey(ses, sshca_key_id)

    request = open("req.dat", "rb").read()
    cert = key.sign_ssh_certificate(sshca_template_id, request, ALGORITHM.RSA_PKCS1_SHA256)

certify()
```
It yields the following exception:

```
$ ./certify.py
Traceback (most recent call last):
  File "/home/mkarwat/repos/ssh-certify/./certify.py", line 25, in <module>
    certify()
  File "/home/mkarwat/repos/ssh-certify/./certify.py", line 23, in certify
    cert = key.sign_ssh_certificate(sshca_template_id, request, ALGORITHM.RSA_PKCS1_SHA256)
  File "/home/mkarwat/.local/lib/python3.9/site-packages/yubihsm/objects.py", line 715, in sign_ssh_certificate
    return self.session.send_secure_cmd(COMMAND.SIGN_SSH_CERTIFICATE, msg)
  File "/home/mkarwat/.local/lib/python3.9/site-packages/yubihsm/core.py", line 372, in send_secure_cmd
    return _unpad_resp(self._secure_transceive(msg), cmd)
  File "/home/mkarwat/.local/lib/python3.9/site-packages/yubihsm/core.py", line 71, in _unpad_resp
    raise YubiHsmDeviceError(six.indexbytes(resp, 3))
yubihsm.exceptions.YubiHsmDeviceError: INVALID_DATA (error code 0x02)

```
The very same request (req.dat) invoked via yubihsm-shell works a treat:

`yubihsm-shell -Chttp://***:12347 --password=*** --authkey=0x4001 --object-id=0x4012 --action=sign-ssh-certificate  --template-id=0x4011 --algorithm=rsa-pkcs1-sha256 --in=req.dat --out=cert.pub`

The library version is 2.1.0.
I would greatly appreciate any hints.

**Comment by trebortech** [comment]
@peyr0l
What dates do you have set in the req? not_before and not_after?

Disclaimer, I haven't yet created the signing path with python-yubihsm, working on completing that today. Will let you know if I am successful or not.

**Comment by trebortech** [comment]
I was able to get past the INVALID_DATA error with the following update

msg = struct.pack("!HHb", self.id, template_id, algorithm.value) + request

Now I'm getting a  SSH_CA_CONSTRAINT_VIOLATION but that is most likely due to something I messed up during testing on the object capabilities.

**Comment by trebortech** [comment]
The sign_ssh_certificate method seems to be missing some modification to the data that is returned from the secure_cmd call. 

The _yubihsm-shell/src/commands.c_  completes some formatting of the returned data that does not look to be in the python version.

https://github.com/Yubico/yubihsm-shell/blob/4d6b1acce356787cf91ddbf355b3b6a25c451c0b/src/commands.c#L2536

I'm working on translating this to python but it could take me some time as my C understanding is limited.

**Comment by trebortech** [comment]
Here is the update I'm testing currently.

You will need to import the b64decode function

```
from base64 import b64decode
```

```
    def sign_ssh_certificate(
        self, template_id, request, algorithm=ALGORITHM.RSA_PKCS1_SHA1
    ):
        """Sign an SSH certificate request.

        :param int template_id: The ID of the SSH TEMPLATE to use.
        :param bytes request: The SSH certificate request.
        :return: The signed SSH certificate.
        :rtype: bytes
        """
        msg = struct.pack("!HHB", self.id, template_id, algorithm.value) + request
        sig = self.session.send_secure_cmd(COMMAND.SIGN_SSH_CERTIFICATE, msg)
        sshcert = b"ssh-rsa-cert-v01@openssh.com "
        sshcert += b64encode(sig)
        sshcert += b"\n"
        return sshcert

```

**Comment by peyr0l** [comment]
Thank you very much for your support!
I will try the patch you proposed today and let you know how well it worked.

**Comment by dainnilsson** [comment]
Thanks for the report and the investigation! 

The missing `algorithm` is indeed a bug in the library and we'll make sure to fix that!

The data returned from the `sign_ssh_certificate` is the binary certificate. To be used by OpenSSH you'll need to convert it into a suitable format. I'm not very knowledgeable in this format, but here's what I've gathered:

The file should start with a fixed string containing with the certificate type, this is hardcoded in yubihsm-shell to be `ssh-rsa-cert-v01@openssh.com`. Then follow a space (" "), then the base64 encoded certificate. This is optionally followed by another space and then a comment. You should quite easily be able to replicate this in Python using something like:


```py
cert = key.sign_ssh_certificate(sshca_template_id, request, ALGORITHM.RSA_PKCS1_SHA256)
with open('cert.pub', 'wb') as f:
    f.write("ssh-rsa-cert-v01@openssh.com " + b64encode(cert) + " comment_here")
```

**Comment by trebortech** [comment]
Thanks @dainnilsson 
I'll give it a try.

btw, I found this helpful on the format

Line 78 for reference on format.
https://cvsweb.openbsd.org/src/usr.bin/ssh/PROTOCOL.certkeys?annotate=HEAD

**Comment by dainnilsson** [comment]
Sorry, I just realized I was mistaken. The response returned is the signature only, not the full signed certificate. As I understand it (not tested) the signature is just appended at the end of the certificate request, so the code would instead be:

```py
signature = key.sign_ssh_certificate(sshca_template_id, request, ALGORITHM.RSA_PKCS1_SHA256)
with open('cert.pub', 'wb') as f:
    f.write(b"ssh-rsa-cert-v01@openssh.com " + b64encode(request + signature) + b" comment_here")
```

**Comment by trebortech** [comment]
I still think we are missing something. The file is a bit long compared to others I have working from using yubihsm-shell.

I'm getting invalid key: invalid format

Here is the formatting string from the example code. I just haven't had any luck trying to understand what the 4 and 256 are. Not sure if they are bytes added or just adding 260 value to the request.

BIO_write(bio, ssh_req + 4 + 256, ssh_req_len + ssh_cert_len - 4 - 256);
https://github.com/Yubico/yubihsm-shell/blob/4d6b1acce356787cf91ddbf355b3b6a25c451c0b/examples/ssh.c#L288

**Comment by dainnilsson** [comment]
If you're able to share samples of a known good certificate (created using yubihsm-shell) and a "bad" one from python-yubihsm, I could take a deeper look!

**Comment by trebortech** [comment]
Here is the working cert

```
ssh-rsa-cert-v01@openssh.com AAAAHHNzaC1yc2EtY2VydC12MDFAb3BlbnNzaC5jb20AAAAg5OpzL6ZFzHQdGB7u6y0Qw0dTfKgCs6tK/qsLGO9QnvsAAAADAQABAAABgQCW1VXaz5lB76l4LhJo6mmDhj/JYHl/nItbrbbfGc5Zrur/hgpZW7fHTvUFTjHvsG8IcDrhELAYrB9OkGhKGhXkeTEYk5VIdrZK1I2P4YOpPEVapRg1ceZgVNdoGeJ4B7z64Emvjzc3zbgiRa2bQhLmrTxIobM7aawPy9vyC1RXtNiD8dKxL4x3IFhPryMy6Z0qbduVpBWkkob4Is0U2iUtOKTmNieKPex+VhTZbSqYURaNdgAEN9DrtRWM6wm9vHjxxCgcA5ONvIAmKzOwTFDq9f/XpLuHd/8LHCG2Iskhv3rWHM7if/W5sV0bNYNxB2DhfEYrzKeR0rmvhofOOyPmK1VhnzKRcYfCCdL3UD+LpDMpztiJIxn5xdCfLR9lpfmqiaMOjCs35Q7mnU7gxBxk3N8v/MvUrPD1kel3FM7dbZ03gVMH0NJarm9hM8o3paKr+a9mY3jI/fVZtNyTv7mTrhF5Za+9xfqqNl6N3fYRgkp0yS73GmXHpZFFTnVoZbMAAAAAAAAAAAAAAAEAAAANdXNlci1pZGVudGl0eQAAAAoAAAAGcmJvb3RoAAAAAGMaHiwAAAAAYxqqzAAAAAAAAACCAAAAFXBlcm1pdC1YMTEtZm9yd2FyZGluZwAAAAAAAAAXcGVybWl0LWFnZW50LWZvcndhcmRpbmcAAAAAAAAAFnBlcm1pdC1wb3J0LWZvcndhcmRpbmcAAAAAAAAACnBlcm1pdC1wdHkAAAAAAAAADnBlcm1pdC11c2VyLXJjAAAAAAAAAAAAAAEXAAAAB3NzaC1yc2EAAAADAQABAAABAQC1lQbeEW0J6uMmsEknuRhG+nrYbv2LEAubkJgHGOkKjsWfFCMCr0D0t1xgSwHLZVKokLSnMeltQFkHWizwFPVi2TnxsU7cygWTpsooswzoIN6Rp6f5LhHYEcOrwy9HEOgcYD305MWH0oQub5jSOytrKi0LKeMiPF4d2BsRoGV1u2+nCq1pGQXtsoCPF8hV1U/rMudtwz8nvVG3G+2qsVO2AfvL8jAdr/G8lYIjSRdQgqD+sHZC7z+BpSMYYJwLgNgUKD+pl9n3H5EDYT+mg/VKyoEVE5cQGXx01ahXIisga78QiSykYHDOSAuwridXJOT4FArS4pljXQrtd/wOklqjAAABFAAAAAxyc2Etc2hhMi0yNTYAAAEAmTlLSOUXz/wN9uL2cK7N3nTQjWRFt4FwOjxDE2MxoKyy7maiujlWHzvXxc8rSCz1KzMNjscsA1WUfY5CPaQuYM7HgChM/QUc3Y59HODf72vX9CE7c/4GFmrZEy5g6sUM/0CV9IyGDvLY87bry1TnasiXqs8qOnHrgv2EOyzC1ZiTlzySQgI36xTETAbzDR9hnLnTi+mWC9M/qyVrprAQGNmgNEIgIFc6A/l78gtpwiEh3m42zyjqfr4l0INRXTH7QKT1vmATLkXBZzE8ny2ntoFgbs2AO34yBSI86Rj1z3eBZrYLkPhp287fI4SyAbqtvJ5bu1wEeP61eHoAGnUOkg==
```

This was is returning invalid format.

```
ssh-rsa-cert-v01@openssh.com YzCvhEkDltcC0vSAcC1NuY9ZNWx4ha+CA4+z2JryLuYjKQEux9RLUXlt/mZKZcpl9o8lTDsJ4Tdvl8Rs/3aoYr5lVAFLZxOsQht8fzf7DIrR6Su37VesB1hh0OPwQRBkKW3qFJ6TpS08joEGnLj/P2hNd4EUF400M1X4qh7zY6dT0YIwpRkv9UjxcJ+FcvgHavLUb6yS4brvhAqUKO5GGYDYAjzQwA4q3Hy68oan9cfSPQ3Li+mkSWocpZy3JLc8zv9RNkWyOYsmvzW5MlE9hlqzGnejN/XAspl9JvJl0cotZUdCMiyM5P8eAnqWLOwAcBOI3LOHOHDul9Sh3sMTmg8wVugAAAAcc3NoLXJzYS1jZXJ0LXYwMUBvcGVuc3NoLmNvbQAAACD5xGLdAjXrgeWhSxlBxc6xvjkvz26kxrtIq3iTPPfGFAAAAAMBAAEAAAGBAJbVVdrPmUHvqXguEmjqaYOGP8lgeX+ci1uttt8Zzlmu6v+GCllbt8dO9QVOMe+wbwhwOuEQsBisH06QaEoaFeR5MRiTlUh2tkrUjY/hg6k8RVqlGDVx5mBU12gZ4ngHvPrgSa+PNzfNuCJFrZtCEuatPEihsztprA/L2/ILVFe02IPx0rEvjHcgWE+vIzLpnSpt25WkFaSShvgizRTaJS04pOY2J4o97H5WFNltKphRFo12AAQ30Ou1FYzrCb28ePHEKBwDk428gCYrM7BMUOr1/9eku4d3/wscIbYiySG/etYczuJ/9bmxXRs1g3EHYOF8RivMp5HSua+Gh847I+YrVWGfMpFxh8IJ0vdQP4ukMynO2IkjGfnF0J8tH2Wl+aqJow6MKzflDuadTuDEHGTc3y/8y9Ss8PWR6XcUzt1tnTeBUwfQ0lqub2Ezyjeloqv5r2ZjeMj99Vm03JO/uZOuEXllr73F+qo2Xo3d9hGCSnTJLvcaZcelkUVOdWhlswAAAABjMK+EAAAAAQAAAA11c2VyLWlkZW50aXR5AAAAHgAAAAFyAAAAAWIAAAABbwAAAAFvAAAAAXQAAAABaAAAAABlEeMEAAAAAGMwr0gAAAAAAAAAggAAABVwZXJtaXQtWDExLWZvcndhcmRpbmcAAAAAAAAAF3Blcm1pdC1hZ2VudC1mb3J3YXJkaW5nAAAAAAAAABZwZXJtaXQtcG9ydC1mb3J3YXJkaW5nAAAAAAAAAApwZXJtaXQtcHR5AAAAAAAAAA5wZXJtaXQtdXNlci1yYwAAAAAAAAAAAAABFwAAAAdzc2gtcnNhAAAAAwEAAQAAAQEAqCEqUvbkrZw3Zz308FNgYs2FKAv4jdmWhYoj6+d3Y/FSv5jtxlXF5mD7qNhTVijrdSuQomQIIiYaSgrokUyomETxUrdt0tZCmnFMpNsuXB0tUvsMCQJW/B7fjvfpesfR/Z1F4zNQBLL9LU5aNF+a37Nhoh9Tf8PBhDZi2F6IWmXY7PH1WFpCGW55Fb12C39ywEt4tPaBGmdXeSJY3Yxjirl7Kn479veDzwXhl/WbHz5vxfAu6W3U6QNWTRuY3GPVKHvLkX4A39SWqtqa7ZjCMNdnMO/M9V1D3ywDJS46gYVvYOhOPwfPiuNKKanoE1CqZw3s+S0C7ijOkMJN2mefJQAAAQ8AAAAHc3NoLXJzYQAAAQCUiyNucGfpEgk37Zz339cujtApO8y6+T6oEbU0fMCRBixOFqWRWps0RP9Nu3Wwbmvk/ta6SX96AvXAVbxzsZj5vFQciAEYbQ0IlSG7INe/8/L/wa221MlS8d9blem8NOLDIRCxpF4BX3+oX9MzORvivjIiq9isJ6g4OW1gwQGnBvXcYluakrZbmRkIIDYlTPIy660rZK07eQOBJ6y5scH/Pqj8B94Vvae/n6Nvxs+UzbtieBd99k2H5g3GPAIthpfvfPzwvRENRDrDpKI9mdMvClskUgUNfuEsI8soTx2/KZma5BZWOVctnKGmDzWxqINgLFGST77jizmd0RIIPeS/ my test
```

**Comment by dainnilsson** [comment]
Ah, I think I see what the issue is. The `request` contains the signed timestamp (4 byte timestamp, 256 byte signature), which shouldn't be part of the certificate output. That needs to be stripped from the `request`:

```py
f.write(b"ssh-rsa-cert-v01@openssh.com " + b64encode(request[4 + 256:] + signature) + b" comment_here")
```

**Comment by trebortech** [comment]
BAM! That did it!!!

Thanks for the help on this.

**Comment by dainnilsson** [comment]
The behavior of `sign_ssh_certificate` may be reconsidered for the next major version of the library, as it is backwards breaking change. The fix can go out in a patch release. Let's leave this issue open until the fix has been merged, I can do that.

**Comment by trebortech** [comment]
I would prefer that the certificate formatting happens outside of python-yubihsm as long as documentation is provided on d.y.c. on how to do this. I made updates to the method return description since that was causing most of my problems.

Please consider that you have to create the template outside of this library so I think it makes sense to also create the certificate with the supplied signature. It will also allow me and others to quickly implement new SSH certificate formats as they become available. My next step is to see if I can get the *-sk keys to work.

**Comment by dainnilsson** [comment]
This has now been released in 2.1.2.
